File: src/problems_all/303_range_sum_query_immutable/Go/v2/303_range_sum_query_immutable_v2.go
/*
https://leetcode.com/problems/range-sum-query-immutable/description/

Runtime: 90 ms
Memory: 9.5 MB
https://leetcode.com/problems/range-sum-query-immutable/submissions/881291377/
*/

package main

import "fmt"

type NumArray struct {
	prefixSum []int
}

func Constructor(nums []int) NumArray {
	prefixSum := make([]int, len(nums)+1)
	for i := 0; i < len(nums); i++ {
		prefixSum[i+1] = prefixSum[i] + nums[i]
	}

	return NumArray{prefixSum: prefixSum}
}

func (this *NumArray) SumRange(left int, right int) int {
	return this.prefixSum[right+1] - this.prefixSum[left]
}

func main() {
	nums := []int{1, 2, 3, 4, 5}
	arr := Constructor(nums)

	res := arr.SumRange(0, 4)
	fmt.Println(res)
}

File: src/problems_all/303_range_sum_query_immutable/303_range_sum_query_immutable_v2.py
"""
https://leetcode.com/problems/range-sum-query-immutable/description/

Runtime:    62 ms
Memory:     20.2 MB
https://leetcode.com/problems/range-sum-query-immutable/submissions/1347037242/
"""

from typing import List


class NumArray:

    def __init__(self, nums: List[int]):
        self.prefix_sum = [0] * (len(nums) + 1)
        for i in range(len(nums)):
            self.prefix_sum[i + 1] = self.prefix_sum[i] + nums[i]

    def sumRange(self, left: int, right: int) -> int:
        return self.prefix_sum[right + 1] - self.prefix_sum[left]


if __name__ == "__main__":
    nums = [1, 2, 3, 4, 5]
    arr = NumArray(nums)
    res = arr.sumRange(0, 4)
    print(res)
